### PR TITLE
BUG: Negative freq in date_range produces values out of start and endpoints

### PR DESCRIPTION
**Pull request by jrmylow**
- [X] closes https://github.com/pandas-dev/pandas/issues/56147
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jrmylow** [comment]
pre-commit.ci autofix

**Comment by MarcoGorelli** [comment]
thanks, taking a look next week - could you fix up the conflicts please?

**Comment by jrmylow** [comment]
pre-commit.ci autofix

**Comment by jrmylow** [comment]
Awaiting https://github.com/pandas-dev/pandas/pull/56873 to fix the test failure in:
```
__________________ test_api_categorical[postgresql_adbc_conn] __________________
[gw0] linux -- Python 3.9.18 /home/runner/micromamba/envs/test/bin/python3.9
[XPASS(strict)] categorical dtype not implemented for ADBC postgres driver
```

**Comment by MarcoGorelli** [comment]
this is probably fine, I just haven't had a chance yet to check how (in?)consistently negative ranges are currently handled. I'm aiming to get this in by 3.0 if it's ready

**Review by MarcoGorelli — COMMENTED** [review]
I think you're right, this is the fix

could you add a whatsnew note for 3.0 please?

could you also update the `date_range` docs please? they currently don't account for negative frequencies, they just say

> Returns the range of equally spaced time points (where the difference between any two adjacent points is specified by the given frequency) such that they all satisfy start <[=] x <[=] end,

If we're changing negative freq behaviour, we need to be very explicit on what the rules are

**Comment by jrmylow** [comment]
pre-commit.ci autofix

**Comment by jrmylow** [comment]
pre-commit.ci autofix

**Review by MarcoGorelli — APPROVED** [review]
thanks @jrmylow , CI failure is unrelated

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -274,7 +274,8 @@ Categorical
 
 Datetimelike
 ^^^^^^^^^^^^
-- Bug in :func:`date_range` where the last valid timestamp would sometimes not be produced (:issue:`56134`)
+- Fixed bug in :func:`date_range` where the last valid timestamp would sometimes not be produced (:issue:`56134`)
+- Fixed bug in :func:`date_range` where using a negative frequency value would not include all points between the start and end values (:issue:`56382`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:278]:
```suggestion
- Bug in :func:`date_range` where the last valid timestamp would sometimes not be produced (:issue:`56134`)
- Bug in :func:`date_range` where using a negative frequency value would not include all points between the start and end values (:issue:`56382`)
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jrmylow

**Pull request merged**